draft-irtf-cfrg-hash-to-curve.md

```diff
@@ -1581,75 +1568,138 @@ Steps:
-twisted Edwards curve is birationally equivalent to a Montgomery curve
-({{BBJLP08}}, Theorem 3.2). To hash to a twisted Edwards curve, hash
-to the equivalent Montgomery curve and evaluate the rational map to
-obtain a point on the twisted Edwards curve.
-
+## Mappings for Twisted Edwards curves
+
+Twisted Edwards curves
+(a class of curves that includes "untwisted" Edwards curves)
+are closely related to Montgomery
+curves ({{montgomery}}): every twisted Edwards curve is birationally equivalent
+to a Montgomery curve ({{BBJLP08}}, Theorem 3.2).
+This equivalence yields an efficient way of hashing to a twisted Edwards curve:
+first, hash to the equivalent Montgomery curve, then transform the
+result into a point on the twisted Edwards curve via a birational map.
+This method of hashing to a twisted Edwards curve thus requires identifying a
+corresponding Montgomery curve and birational map.
+We describe how to identify such a curve and map immediately below.
+
+### Birational maps from Montgomery to Twisted Edwards curves {#birational-map}
+
+There are two ways to identify the correct Montgomery curve and
+birational map for use when hashing to a given twisted Edwards curve.
+
+When hashing to a standardized twisted Edwards curve for which a corresponding
+Montgomery form and birational map are also standardized, the standard
+Montgomery form and birational map MUST be used to ensure compatibility
+with existing software.
+Two such standardized curves are the edwards25519 and edwards448 curves,
+which correspond to the Montgomery curves curve25519 and curve448, respectively.
+For both of these curves, {{RFC7748}} lists both the Montgomery and twisted Edwards
+forms and gives the corresponding rational maps.
+
+The birational map for edwards25519 ({{RFC7748}}, Section 4.1)
+uses the constant sqrt\_neg\_486664 = sqrt(-486664) mod 2^255 - 19.
+To ensure compatibility, this constant MUST be chosen such that
+sgn0(sqrt\_neg\_486664) == 1.
+Analogous ambiguities in other standardized birational maps MUST be
+resolved in the same way: for any constant k whose sign is ambiguous,
+k MUST be chosen such that sgn0(k) == 1.
+
+The 4-isogeny map from curve448 to edwards448 ({{RFC7748}}, Section 4.2)
+is unambiguous with respect to sign.
+
+When defining new twisted Edwards curves, a Montgomery equivalent and birational
+map SHOULD be specified, and the sign of the birational map SHOULD be stated
+unambiguously.
+
+When hashing to a twisted Edwards curve that does not have a standardized
+Montgomery form or birational map, the following procedure MUST be
+used to derive them.
 For a twisted Edwards curve given by a * x^2 + y^2 = 1 + d * x^2 * y^2,
-first compute A and B, the parameters of the equivalent Montgomery curve,
-as follows:
+first compute A and B, the parameters of the equivalent Montgomery curve
+y'^2 = x'^3 + A * x'^2 + B * x', as follows:
 
-- A = (a + 2) / 2
+- A = (a + d) / 2
```

Comment:
Question: how did you get B and B'? _or equivalently _ how do you convert between 
```
y^2 = x^3 + A * x^2 + C * x
```
and
```
B y^2 = x^3 + A * x^2 +  x
```
?

Reply:
See c56449fc41f0e0fdebfa3bd4e8e7e83cb6027a80

Reply:
Also note: I implemented the described birational map for edwards448 in the `poc` directory. The map correctly transports points between edwards448 and an equivalent Montgomery curve (though not the one specified in RFC7748; that's a TODO for code update)

Reply:
I think the conversation in #134 covers this now, right @armfazh?